server/src/main/kotlin/suwayomi/tachidesk/server/Settings.kt

```diff
@@ -0,0 +1,233 @@
+            val settingInDb = settingsInDb[it.name]
+            val isSettingMissing = settingInDb == null
+            val updateRequired =
+                setting.value.toString() != settingInDb?.value ||
+                    setting.requiresRestart != null && setting.requiresRestart != settingInDb?.requiresRestart
+
+            val shouldDoUpdate = !isSettingMissing && doUpdate && updateRequired
+
+            if (shouldDoUpdate || isSettingMissing) {
+                KotlinLogging.logger { }
+                    .debug { "Setting \"${it.name}\" - insert= $isSettingMissing, update= $shouldDoUpdate - $setting" }
+            }
+
+            if (shouldDoUpdate) {
+                put(it.name, setting)
+                return@forEach
+            }
+
+            if (!isSettingMissing) {
+                return@forEach
+            }
+
+            transaction {
+                SettingsTable.insert { table ->
+                    table[key] = it.name
+                    table[value] = setting.value.toString()
+                    if (setting.requiresRestart != null) table[requiresRestart] = setting.requiresRestart!!
+                }
+            }
+        }
+
+        updateCache()
+    }
+
+    fun reset() {
+        insertAndOrUpdate(DEFAULT_SETTINGS, true)
+    }
+
+    fun setup() {
+        insertAndOrUpdate(DEFAULT_SETTINGS, false)
+    }
+
+    fun update(settings: IServerSettings) {
+        insertAndOrUpdate(settings, true)
+    }
+
+    fun fromConfig(config: ServerConfig): IServerSettings {
+        return PartialServerSettings(
+            ip = PartialSetting(value = config.ip),
+            port = PartialSetting(value = config.port),
+            socksProxyEnabled = PartialSetting(value = config.socksProxyEnabled),
+            socksProxyHost = PartialSetting(value = config.socksProxyHost),
+            socksProxyPort = PartialSetting(value = config.socksProxyPort),
+            webUIEnabled = PartialSetting(value = config.webUIEnabled),
+            webUIFlavor = PartialSetting(value = config.webUIFlavor),
+            initialOpenInBrowserEnabled = PartialSetting(value = config.initialOpenInBrowserEnabled),
+            webUIInterface = PartialSetting(value = config.webUIInterface),
+            electronPath = PartialSetting(value = config.electronPath),
+            downloadAsCbz = PartialSetting(value = config.downloadAsCbz),
+            downloadsPath = PartialSetting(value = config.downloadsPath),
+            maxParallelUpdateRequests = PartialSetting(value = config.maxParallelUpdateRequests),
+            basicAuthEnabled = PartialSetting(value = config.basicAuthEnabled),
+            basicAuthUsername = PartialSetting(value = config.basicAuthUsername),
+            basicAuthPassword = PartialSetting(value = config.basicAuthPassword),
+            debugLogsEnabled = PartialSetting(value = config.debugLogsEnabled),
+            systemTrayEnabled = PartialSetting(value = config.systemTrayEnabled)
```

Comment:
the idea would be to completely remove the old config file and only use the settings from the database.
the settings would then only be changeable via the UI (and via a launcher, in case that is still planned?)

regarding the code duplication, this is so that there is a partial data class of the settings, so that via the endpoint a single setting can be changed, without having to send all settings
not sure if this can be done in a better way, haven't found anything besides this

so regarding the backend of the settings, I understand you still want to have a config file?
if java.util.prefs.Preferences should be used, if I understand it correctly, we have to implement that ourself, since by default it will be saved in the registry

Reply:
> the idea would be to completely remove the old config file and only use the settings from the database.
the settings would then only be changeable via the UI (and via a launcher, in case that is still planned?)

I don't really like that idea because we have a lot of headless (docker, NAS, etc) users that prefer the sys admin conventional way, we also service home users, we might want to make things easier and more simple stupid for them but we shouldn't turn our back to those power users. 

> regarding the code duplication, this is so that there is a partial data class of the settings, so that via the endpoint a single setting can be changed, without having to send all settings
not sure if this can be done in a better way, haven't found anything besides this

@Syer10 any ideas?

I could try to come up with something when I have the time.

> so regarding the backend of the settings, I understand you want to keep a config file?

Yes, as explained before.

> if java.util.prefs.Preferences should be used, if I understand it correctly, we have to implement that ourself, since by default it will be saved in the registry

About that one I was just throwing the an idea, now that I think about it it might not be the best idea? 
- Java prefs are something that is kept hidden away and it's not an industry standard (besides in android) 
- Keeping the config in cleartext and accessible is almost a must, even though one of our regulars on discord likes to play with the database file but that would be annoying to anyone else.